### ISSUE TITLE
The next.js option transpilePackages does not transpile inherited packages with Turbopack (--turbo)

### ISSUE THREAD

**Issue by ChiefORZ** [vercel/next.js#63230]
### Link to the code that reproduces this issue

https://github.com/ChiefORZ/next-turbo-transpile-packages

### To Reproduce

1. Start the application in development (pnpm run dev)
2. You will see that everything works fine and gets transpiled
3. Start the application in development with turbo (pnpm run dev:turbo)
4. You will see the error `Module not found: Can't resolve '@example/lib-a'`

### Current vs. Expected behavior

Following the steps in the previous section, i expected the turbo dev server (step 4) to transpile the packages the same way as before (step 2)

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.3.0: Wed Dec 20 21:33:31 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T8112
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: N/A
  pnpm: 8.15.1
Relevant Packages:
  next: 14.1.3
  eslint-config-next: 14.1.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack (--turbo)

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

In our monorepo setup we have a couple of libraries that are shared across multiple apps. Because the libraries are only used from apps within this repository we don't need to compile them and release them to a node registry. Therefore the next.js config option **transpilePackages** is a wonderful gift 🎉
In these shared libraries we depend on other libraries within this repo, everything works fine.
But we tried to switch to Turborepo and the only blocker is the **transpilePackages** option for us, we tried to precompile the pkgs and `--turbo` works flawlessly without **transpilePackages**.

Remarks: When creating this example repository i saw that the **transpilePackages** option works if you try to import a lib that does not have a dependency to another lib (`@example/lib-a`). 

<sub>[PACK-2734](https://linear.app/vercel/issue/PACK-2734/the-nextjs-option-transpilepackages-does-not-transpile-inherited)</sub>

**Comment by bgw** [comment]
Hey @ChiefORZ. Thanks for taking the time to try out turbopack, to provide feedback about your experience, and to create a detailed repro repository.

I was able to reproduce your issue with `pnpm run dev:turbo`, but I also ran into similar issues without turbo, so I think there’s a broader configuration issue with your test repository.

After cloning your repository and running `pnpm i`, I tried running `pnpm run dev` *(without turbo!)* as suggested in steps 1 and 2. Loading the development server in the browser shows the following error modal:

![Screenshot 2024-04-08 at 3 58 41 PM](https://github.com/vercel/next.js/assets/180404/71f76989-9b06-4272-a337-e600c2d23d17)

It looks like `app/tsconfig.json` is read by the next.js dev server, but `lib-a/tsconfig.json` and `lib-b/tsconfig.json` are unused.

While it seems like it might be nice for next.js to also read those `tsconfig.json` files and use them for module resolution, the current behavior seems to match the behavior of `tsc`. When running `cd app && node_modules/.bin/tsc`, I get the following similar errors:

```
src/app/page.tsx:1:10 - error TS2305: Module '"@example/lib-b"' has no exported member 'generateInvalidMsg'.

1 import { generateInvalidMsg } from "@example/lib-b";
           ~~~~~~~~~~~~~~~~~~

../libs/lib-b/src/index.ts:1:15 - error TS2307: Cannot find module '@/utils' or its corresponding type declarations.

1 export * from '@/utils';
                ~~~~~~~~~


Found 2 errors in 2 files.

Errors  Files
     1  src/app/page.tsx:1
     1  ../libs/lib-b/src/index.ts:1
```
